hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -1052,3 +959,206 @@ def array_for(index_shape, size):
+
+        elements = cu.many(
+            data,
+            min_size=min_total,
+            max_size=max_total,
+            average_size=min(
+                max(min_total * 2, min_total + 5), 0.5 * (min_total + max_total)
+            ),
+        )
+
+        # All shapes are handled in column-major order; i.e. they are reversed
+        shapes = [[] for _ in range(self.num_shapes)]
+        result_shape = list(self.base_shape[::-1])
+        while elements.more():
+            updated_shape = self._pick_shape(data, shapes)
+            if len(updated_shape) < len(result_shape):
+                # Shrinks towards original base-shape
+                if result_shape[len(updated_shape)] == 1:
+                    if self.min_side <= 1 and not data.draw(st.booleans()):
+                        side = 1
+                    else:
+                        side = data.draw(self.side_strat)
+                elif self.max_side >= result_shape[len(updated_shape)] and (
+                    not self.min_side <= 1 <= self.max_side or data.draw(st.booleans())
+                ):
+                    side = result_shape[len(updated_shape)]
+                else:
+                    side = 1
+            else:
+                side = data.draw(self.side_strat)
+            updated_shape.append(side)
+
+            # update the resulting shape, based on broadcasting semantics
+            if len(updated_shape) > len(result_shape):
+                result_shape.append(updated_shape[-1])
+            elif updated_shape[-1] != 1 and result_shape[len(updated_shape) - 1] == 1:
+                result_shape[len(updated_shape) - 1] = updated_shape[-1]
+
+        max_len = max(len(s) for s in shapes) if shapes else 0
+        result_shape = result_shape[: max(len(self.base_shape), max_len)]
+
+        assert len(shapes) == self.num_shapes
+        assert all(self.min_dims <= len(s) <= self.max_dims for s in shapes)
+        assert all(self.min_side <= s <= self.max_side for side in shapes for s in side)
+
+        return (
+            tuple(tuple(reversed(shape)) for shape in shapes),
+            tuple(reversed(result_shape)),
+        )
+
+
+@st.defines_strategy
+@reserved_means_kwonly_star
+def multiple_shapes(
+    inputs,
+    base_shape=(),
+    __reserved=not_set,
+    min_dims=0,
+    max_dims=None,
+    min_side=1,
+    max_side=None,
+):
+    # type: (int, Shape, Any, int, int, int, int) -> st.SearchStrategy[Tuple[Tuple[Shape, ...], Shape]]
+    """Return a strategy for generating a specified number of shapes, N, that are
+    mutually-broadcastable with one another and with the provided "base-shape".
+
+    The strategy will generate a tuple of:
+     * the N generated shapes
+     * the resulting shape, produced by broadcasting the N shapes with the base-shape
+
+    Each shape produced from this strategy shrinks towards a shape with length
+    ``min_dims``. The size of an aligned dimension shrinks towards being a singleton.
```

Comment:
"singleton" has a particular defined meaning in Python (special objects like `None`), so I'd explicitly say "shrinks towards size one".